docs/NEO-EDEN.md

```diff
@@ -0,0 +1,199 @@
+
+**But how do you debug that test?** Because you're running an interpreter, which runs bash commands via os.exec plus you execute golang program, which is written and compiled inside the repository. Debug printing would work, but I find debuggers much more useful and efficient most of the times. It is a matter of taste, but better to debugger at your disposal, than not to have it in this case. Even worse, how do you debug escript problem? You need to know it's internals, there's no Stackoverflow or Google by your side, nor there are books about it.
+
+**Bumping golang is actually manual action**: last but not least, we have to maintain custom test files which a basically copy-paste with one added function, bumping golang turns into some weird dances in that case.
+
+I believe that those reasons are good enough to switch for golang-only approach to write tests, which is called NeoEden (kudos to @shjala for the name)
+
+Also worst case what could happen with DSL is something like the Inner JSON effect [link](https://thedailywtf.com/articles/the-inner-json-effect)
+
+## Enter NeoEden. How tests look now?
+
+They look like [this](../tests/sec/sec_test.go) code listing is in the end of this section. Conceptually, Eden contains of two entities: an entity which allows you setup some kind of environment and deploy EVE there. Think Infrastructure as a Code, like Teraform, which allows me using golang code to deploy, this is openevec package; second entity is evetestkit, which serves as an abstraction over missing features of adam as controller such as keeping track of state and provide you with toolkit to test things, namely send commands and access EVE or its AppInstances to see if the certain file exists or certain request gives you certain response. Those two things let you create golang-only test like one below. You use openevec to setup environment and evetestkit to help you describe what is passing criterion for particular test. The common denominator between openevec and evetestkit is EdenSetupArgs, openevec is just a wrapper over that configuration, which defines your state. The goal is to be able to define your state, write custom functions to state, like this:
+
+
+```go
+
+pe := openevec.GetDefaultPatchEnvelope()
+appInstance := openevec.GetDefaultAppInstance()
+
+cfg := openevec.GetDefaultConfig(path)
+  .WithHyperVisor(openevec::HyperVisorQemu)
+  .WithAppInstances(appInstance)
+  .WithPatchEnvelopes(pe)
+
+evec := openvec.CreateOpnEVEC(cfg)
+
+evec.SetupEden(/* ... */)
+evec.StartEden(/* ... */)
+evec.OnboardEve(/* ... */)
+
+node, err := tk.InitializeTestFromConfig(cfg)
```

Comment:
What is `tk` here? Is this `node` part of `openevec` or `evetestkit`?

Reply:
evetestkit

Reply:
Got it. Should be clearer.